src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -80,48 +82,53 @@ private JCTree invokeOpToJCMethodInvocation(CoreOp.InvokeOp invokeOp) {
         }
         var methodInvocation = treeMaker.App(meth, args.toList());
         if (invokeOp.isVarArgs()) {
-            var lastParam = invokeOp.invokeDescriptor().type().parameterTypes().getLast();
-            if (lastParam instanceof ArrayType varargType) {
-                methodInvocation.varargsElement = typeElementToType(varargType.componentType());
-            } else {
-                Assert.error("Vararg method doesn't have a trailing array type: " + invokeOp.invokeDescriptor());
-            }
+            setVarargs(methodInvocation, invokeOp.invokeDescriptor().type());
+        }
+        if (invokeOp.result().uses().isEmpty()) {
+            return treeMaker.Exec(methodInvocation);
         }
         return methodInvocation;
     }
 
+    void setVarargs(JCExpression tree, FunctionType type) {
+        var lastParam = type.parameterTypes().getLast();
+        if (lastParam instanceof ArrayType varargType) {
+            TreeInfo.setVarargsElement(tree, typeElementToType(varargType.componentType()));
+        } else {
+            Assert.error("Expected trailing array type: " + type);
+        }
+    }
+
     private JCTree opToTree(Op op) {
         JCTree tree = switch (op) {
             case CoreOp.ConstantOp constantOp when constantOp.value() == null ->
                     treeMaker.Literal(TypeTag.BOT, null).setType(syms.botType);
             case CoreOp.ConstantOp constantOp -> treeMaker.Literal(constantOp.value());
             case CoreOp.InvokeOp invokeOp -> invokeOpToJCMethodInvocation(invokeOp);
-            case CoreOp.NewOp newOp -> {
-                if (newOp.resultType() instanceof ArrayType at) {
-                    var elemType = treeMaker.Ident(typeElementToType(at.componentType()).tsym);
-                    var dims = new ListBuffer<JCTree.JCExpression>();
-                    for (int d = 0; d < at.dimensions(); d++) {
-                        dims.add(exprTree(newOp.operands().get(d)));
-                    }
-                    var na = treeMaker.NewArray(elemType, dims.toList(), null);
-                    na.type = typeElementToType(at);
-                    yield na;
+            case CoreOp.NewOp newOp when newOp.resultType() instanceof ArrayType at -> {
```

Comment:
Shouldn't we have a method like `boolean NewOp.isArrayCreation()` ?

Reply:
Yeah - this will probably be helpful -- but I'd prefer that to be addressed separately